### DISCUSSION TITLE
RESTful API와 REST API의 차이점이 무엇일까요?

### DISCUSSION THREAD

**Discussion by BaekYuri** [wanted-9th-3team/frontend-study#12] [프론트 기술 질문 공간]
RESTful API와 REST API의 차이점이 무엇일까요?

**Comment by michoball** [comment]
REST api 는 웹과 같은 시스템에서 데이터를 안전하게 주고 받기 위해 만든 상호간에 합의된 규약 이라고 할 수 있습니다. 

로이 필딩이라는 사람이 박사학위 논문으로 세계 최초로 공개했고 이분이 제시했을 당시 6가지의 아키텍쳐 스타일을 모두 지켜야한다고 했고 이를 다 지킨 rest api 를 restful api 라고 할 수 있습니다. 

각 스타일에 대해서 자세한 내용은 알지 못하지만 대부분의 스타일은 다들 잘 지키지만 Uniform interface 의 제약조건을 잘 지키지 못하고 지키기 어려운 것으로 알고 있습니다. 

하지만 그렇다고 해서 restful 하지 않다고 걱정할 필요가 없는게 우리가 사용하는 웹 브라우져들이 restful하게 잘 지키고 있기 때문에 괜찮은 것으로 알고 있습니다. 

다시말해서 로이 필딩이 제시한 규약 조건중  Uniform interface를 잘 지키느냐 아니냐로  restful과 그냥 rest api 가 차이가 난다고 할 수 있을 것 같습니다.

**Comment by WooGie911** [comment]
위 명훈님의 의견에 보태자면

로이 필딩의 논문에서 6가지 아키택쳐 스타일은 다음과 같습니다.

1. Client-Server : 클라이언트-서버 스타일은 사용자 인터페이스에 대한 관심(concern)을 데이터 저장에 대한 관심으로부터 분리함으로써 클라이언트의 이식성과 서버의 규모확장성을 개선한다.
2. Stateless : 클라이언트와 서버의 통신에는 상태가 없어야한다. 모든 요청은 필요한 모든 정보를 담고 있어야한다. 요청 하나만 봐도 바로 뭔지 알 수 있으므로 가시성이 개선되고, task 실패시 복원이 쉬우므로 신뢰성이 개선되며, 상태를 저장할 필요가 없으므로 규모확장성이 개선된다.
3. Cache : 캐시가 가능해야한다. 즉 모든 서버 응답은 캐시 가능한지 그렇지 아닌지 알 수 있어야한다. 호율, 규모확장성, 사용자 입장에서의 성능이 개선된다.
4. Uniform Interface : 구성요소(클라이언트, 서버 등) 사이의 인터페이스는 균일(uniform)해야한다. 인터페이스를 일반화함으로써, 전체 시스템 아키텍처가 단순해지고, 상호작용의 가시성이 개선되며, 구현과 서비스가 분리되므로 독립적인 진화가 가능해진다. 이 스타일은 다음의 네 제약조건으로 이루어진다.
5. Layered System : 계층(hierarchical layers)으로 구성이 가능해야하며, 각 레이어에 속한 구성요소는 인접하지 않은 레이어의 구성요소를 볼 수 없어야한다.
6. Code-On-Demand (Optional) : Code-On-Demand가 가능해야한다. 서버가 네트워크를 통해 클라이언트에 프로그램을 전달하면 그 프로그램이 클라이언트에서 실행될 수 있어야한다. (Java applet이나 Javascript 같은 것을 말함) 다만 이 제약조건은 필수는 아니다.


이 아키텍쳐 스타일이 반영되어 다음과 같은 특징을 갖는다고 볼 수 있습니다.


REST의 특징

1. Server-Client (서버-클라이언트 구조)
- 자원이 있는 쪽이 Server, 자원을 요청하는 쪽이 Client가 됩니다.
- REST Server는 API를 제공하고 비즈니스 로직 처리 및 저장을 책임지고,
- Client는 사용자 인증이나 context( 세션, 로그인 정보 ) 등을 직접 관리하고 책임집니다.
- 역할을 확실히 구분시킴으로써 서로 간의 의존성을 줄입니다.

2. Stateless (무상태)
- HTTP 프로토콜은 Stateless Protocol이므로 REST 역시 무상태성을 갖습니다.
- Client의 context를 Server에 저장하지 않습니다.
- 즉, 세션과 쿠키와 같은 context 정보를 신경쓰지 않아도 되므로 구현이 단순해집니다.
- Server는 각각의 요청을 완전히 별개의 것으로 인식하고 처리합니다.
- 각 API 서버는 Client의 요청만을 단순 처리합니다.
- 즉, 이전 요청이 다음 요청의 처리에 연관되어서는 안됩니다. ( DB에 의해 바뀌는 것은 허용 )
- Server의 처리 방식에 일관성을 부여하기 때문에 서비스의 자유도가 높아집니다.

3. Cacheable (캐시 처리 기능)
- 웹 표준 HTTP 프로토콜을 그대로 사용하므로 웹에서 사용하는 기존의 인프라를 그대로 활용할 수 있습니다.
- 즉, HTTP가 가진 가장 강력한 특징 중 하나인 캐싱 기능을 적용할 수 있습니다.
- HTTP 프로토콜 표준에서 사용하는 Last-Modified Tag 또는 E-Tag를 이용해 캐싱을 구현합니다.
- 대량의 요청을 효율적으로 처리할 수 있습니다.

4. Layered System (계층 구조)
- Client는 REST API Server만 호출합니다.
- REST Server는 다중 계층으로 구성될 수 있습니다.
- 보안, 로드 밸런싱, 암호화 등을 위한 계층을 추가하여 구조를 변경할 수 있습니다.
- Proxy, Gateway와 같은 네트워크 기반의 중간매체를 사용할 수 있습니다.
- 하지만 Client는 Server와 직접 통신하는지, 중간 서버와 통신하는지는 알 수 없습니다.

5. Uniform Interface (인터페이스 일관성)
- URI로 지정한 Resource에 대한 요청을 통일되고, 한정적으로 수행하는 아키텍처 스타일을 의미합니다.
- HTTP 표준 프로토콜에 따르는 모든 플랫폼에서 사용이 가능하며, Loosely Coupling(느슨한 결함) 형태를 갖습니다. 
- 즉, 특정 언어나 기술에 종속되지 않음

6. Self-Descriptiveness (자체 표현)
- 요청 메시지만 보고도 쉽게 이해할 수 있는 자체 표현 구조로 되어있습니다.



참고 : https://dev-coco.tistory.com/97
참고 : https://blog.npcode.com/2017/03/02/%EB%B0%94%EC%81%9C-%EA%B0%9C%EB%B0%9C%EC%9E%90%EB%93%A4%EC%9D%84-%EC%9C%84%ED%95%9C-rest-%EB%85%BC%EB%AC%B8-%EC%9A%94%EC%95%BD/
